designs/0031-deprecation-schedule.md

```diff
@@ -0,0 +1,254 @@
+2. It has been added to the [documentation on
+   deprecation](https://mc-stan.org/docs/reference-manual/deprecated-features-appendix.html)
+   with the same 3 pieces of information, and additional information as needed
+   (such as any of: reasoning for removal, advantages of new behavior, ongoing
+   related changes, etc).
+3. Examples in the documentation have been updated to remove any examples of the
+   deprecated syntax, except for those referring to its deprecation. This change
+   may lag behind bullets 1 and 2, but must occur before the removal.
+
+Our goal is to only deprecate a feature when doing so provides a tangible
+benefit to users, such as improving error messaging or allowing new features to
+be added to the language while keeping a low maintenance burden. Changes should
+never take users (even those with many or large models) long to adjust to, and
+as few such changes should be made in each version as is reasonable.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Developers
+From a developer standpoint, this policy is rather straightforward. When a
+change is made, it should identify any deprecation candidates: new keywords that
+must be implemented (thus 'deprecating' the use of that word as a identifier),
+new syntax which supersedes prior syntax, (such as `=` or `array[]`), or a
+similar **syntax-level** change. If supporting both the new and old way of doing
+things requires a 'hack' or difficult to maintain code, this is a prime
+candidate for this deprecation-removal process. If a change meaningfully changes
+the semantics of the language, it is not.
+
+>Current examples include the use of 'partial keywords' like offset,
+>which are not reserved words but act like them in some contexts, or multiple
+>syntaxes which lead to parsing difficulties, like the new array syntax. We would
+>like to not only improve the error messaging and implementation of these current
+>features, but also avoid introducing new issues of this variety in the long term.
+
+>Current non-examples include the ODE solver interface changes, which requires
+>changes to the user-defined function defining the ODE.
+
+Once it has been decided to deprecate a feature, the lexer, parser, or
+typechecker (whichever is most apropriate) should be updated to emit a
+compile-time warning notifying the user. This must contain the above 3 criteria:
+1. What is deprecated
+2. How to adjust your code
+3. When you need to adjust it by.
+
+The final bullet point should be decided during the review process with 3
+versions as a minimum. Many changes, such as reserving a new keyword, will not
+require longer periods, but syntax changes which the user may need time to
+adjust to could require longer time frames.
+
+This information must also be added to the documentation in the accompanying PR
+to the change.
+
+Finally, developers may wish to weigh in on how the implementation might look
```

Comment:
We're not talking about what the devs want, but what they're allowed to do, so:

"wish to" -> ""

But don't we just want to say that they must discuss this as part of deprecating a feature?  Or are you talking about the OCaml code instead of the Stan code?  (It always helps to be as specific as possible in specs.)

Also, I think we can give them the latitude to do this under any circumstances, so:

", if they can foresee..." -> "."

Reply:
I meant the OCaml, meaning something like "Once x is reserved as a keyword formally, we can change the parser to behave like y and simplify its behavior".